yay.8
/n'\" t
.TH "YAY" "8" "2018-02-29" "Yay v3\&.460+" "Yay Manual"
.nh
.ad l
.SH "NAME"
yay \- AUR Helper written in go
.SH "SYNOPSIS"
.sp
\fIyay\fR <operation> [options] [targets]
.sp
\fIyay\fR <search pattern>
.SH "DESCRIPTION"
.sp
Yay is a Pacman wrapper with AUR support\&. It passes options to Makepkg and
Pacman after resolving packages to install/upgrade\&.
.sp
This manpage only covers options unique to Yay\&. For other options see
\fBpacman(8)\fR\&.
.SH "YAY OPERATIONS"
.PP
\fB\-Y, --yay\fR
.RS 4
Perform yay specific operations\&. This is the default if no other operation is
selected\&.
.RE
.PP
\fB\-P, --print\fR
.RS 4
Perform yay specific print operations\&.
.RE
.PP
\fB\-G, --getpkgbuild\fR
.RS 4
Downloads PKGBUILD from ABS or AUR\&.
.RE
.PP
If no operation is selected -Y will be assumed\&.
.SH "EXTENDED PACMAN OPERATIONS"
.PP
\fB\-S, -Si, -Ss, -Su\fR
.RS 4
These operations are extended to support both AUR and repo packages\&.
.RE
.PP
\fB\-R\fR
.RS 4
Yay will also remove cached data about devel packages\&.
.RE
.SH "YAY OPTIONS (APPLY TO -Y AND --YAY)"
.PP
\fB<NO OPTION>\fR
.RS 4
Displays a list of packages matching the search terms and prompts the user on
which packages to install (yogurt mode)\&.
.RE
.PP
\fB   \-\-gendb\fR
.RS 4
Generate development package databse\&. Tracks the latest commit for each
development package, when there is a new commit Yay will know to update\&. This
is done per package whenever a package is synced. This option should only be
used when migrating to Yay from another AUR helper.
.RE
.PP
\fB\-c \-\-clean\fR
.RS 4
Remove unneeded dependencies\&.
.RE
.SH "PRINT OPTIONS (APPLY TO -P AND --PRINT)"
\fB\-c \-\-complete\fR
.RS 4
Print a list of all AUR and repo packages\&. This is to allow shell completion
and is not intended to be used directly by the user\&.
.RE
.PP
\fB\-f \-\-fish\fR
.RS 4
During complete adjust the output for the fish shell\&.
.RE
.PP
\fB\-d \-\-defaultconfig\fR
.RS 4
Print default yay configuration\&.
.RE
.PP
\fB\-g \-\-config\fR
.RS 4
Print current yay configuration\&.
.RE
.PP
\fB\-n \-\-numberupgrades\fR
.RS 4
Print number of packages that need to be updated\&. Note this does not preform
a database refresh\&. Run \fByay -Sy\fR Before this for an up to date result\&.
.RE
.PP
\fB\-s \-\-stats\fR
.RS 4
Displays information about installed packages and system health\&. If there are
orphaned, out-of-date or packages that no longer exist on the AUR warnings will
be displayed\&.
.RE
.PP
\fB\-u \-\-upgrades\fR
.RS 4
Print Names of packages that need to be updated\&. Note this does not preform
a database refresh\&. Run \fByay -Sy\fR Before this for an up to date result\&.
.PP
.SH "PERMANENT CONFIGURATION SETTINGS"
.PP
\fB\-\-save\fR
.RS 4
Causes the following options to be saved back to the config file\&. This
provides an easy way to change config options without directly editing the
file\&.
.RE
.PP
\fB\-\-builddir <dir>\fR
.RS 4
Directory to use for Building AUR Packages\&. This directory is also used as
the AUR cache when deciding if Yay should should skip builds\&.
.RE
.PP
\fB\-\-editor <file>\fR
.RS 4
Editor to use when editing PKGBUILDs\&. If this is not set the \fBEDITOR\fR
environment variable will be checked, followed by \fBVISUAL\fR\&. If none of
these are set Yay will prompt the user for an editor\&.
.RE
.PP
\fB\-\-makepkg <file>\fR
.RS 4
The command to use for \fBmakepkg\fR calls. This can be a command in
\fBPATH\fR or an absolute path to the file\&.
.RE
.PP
\fB\-\-pacman <file>\fR
.RS 4
The command to use for \fBpacman\fR calls. This can be a command in
\fBPATH\fR or an absolute path to the file\&.
.RE
.PP
\fB\-\-tar <file>\fR
.RS 4
The command to use for \fBbsdtar\fR calls. This can be a command in
\fBPATH\fR or an absolute path to the file\&.
.RE
.PP
\fB\-\-git <file>\fR
.RS 4
The command to use for \fBgit\fR calls. This can be a command in
\fBPATH\fR or an absolute path to the file\&.
.RE
.PP
\fB\-\-config <file>\fR
.RS 4
The pacman config file to use\&.
.RE
.PP
\fB\-\-requestsplitn <n>\fR
.RS 4
The maximum amount of packages to request per AUR query\&. The higher the
number the faster AUR requests will be\&. Requesting too many packages in one
AUR query will cause an error\%. This should only make a noticeable difference
with very large requests (>500) packages\&.
.RE
.PP
\fB\-\-topdown\fR
.RS 4
Display repository packages first and then AUR packages\&.
.RE
.PP
\fB\-\-bottomup\fR
.RS 4
Show AUR packages first and then repository packages\&.
.RE
.PP
\fB\-\-devel\fR
.RS 4
During sysupgrade also check AUR development packages for updates\&. Currently
only GitHub packages are supported\&.
.RE
.PP
\fB\-\-nodevel\fR
.RS 4
Do not check for development packages updates during sysupgrade\&.
.RE
.PP
\fB\-\-afterclean\fR
.RS 4
Remove package sources after successful Install\&.
.RE
.PP
\fB\-\-noafterclean\fR
.RS 4
Do not remove package sources after successful Install\&.
.RE
.PP
\fB\-\-timeupdate\fR
.RS 4
During sysupgrade also compare the build time of installed packages against
the last modification time of each package's AUR page\&.
.RE
.PP
\fB\-\-notimeupdate\fR
.RS 4
Do not consider build times during sysupgrade\&.
.RE
.PP
\fB\-\-redownload\fR
.RS 4
Always download pkgbuilds of targets even when a copy is available in cache\&.
.RE
.PP
\fB\-\-redownloadall\fR
.RS 4
Always download pkgbuilds of all AUR packages even when a copy is available
in cache\&.
.RE
.PP
\fB\-\-noredownload\fR
.RS 4
When downloading pkgbuilds if the pkgbuild is found in cache and is equal or
newer than the AUR's version use that instead of downloading a new one\&.
.RE
.PP
\fB\-\-rebuild\fR
.RS 4
Always build target packages even when a copy is available in cache\&.
.RE
.PP
\fB\-\-rebuildall\fR
.RS 4
Always build all AUR packages even when a copy is available
in cache\&.
.RE
.PP
\fB\-\-rebuildtree\fR
.RS 4
When installing an AUR package rebuild and reinstall all of its AUR
dependencies recursivley, even the ones already installed. This flag allows
you to easily rebuild packages against your current system's libraries if they
have become incompatible.
.RE
.PP
\fB\-\-norebuild\fR
.RS 4
When building packages if the package is found in cache and is an equal version
to the one wanted skip the package build and use the existing package\&.
.RE
.PP
\fB\-\-mflags <flags>\fR
.RS 4
Passes arguments to makepkg\&. These flags get passed to every instance where
makepkg is called by Yay. Arguments are split on whitespace before being
passed to makepkg. Multiple arguments may be passed by supplying a space
separated list that is quoted by the shell.
.RE
.PP
\fB\-\-sudoloop\fR
.RS 4
Loop sudo calls in the background to prevent sudo from timing out during long
builds\&.
.RE
.PP
\fB\-\-nosudoloop\fR
.RS 4
Do not loop sudo calls in the background\&.
.RE
.SH "EXAMPLES"
.PP
yay \fIfoo\fR
.RS 4
Search and install from the repos and the \fBAUR\fR\ using yogurt mode\&.
.RE
.PP
yay -Syu
.RS 4
Update package list and upgrade all currently installed repo and \fBAUR\fR\&.
.RE
.PP
yay -S \fIfoo\fR
.RS 4
Installs package \fIfoo\fR from the repos or the \fBAUR\fR\&.
.RE
.PP
yay -Ss \fIfoo\fR
.RS 4
Searches for package \fIfoo\fR on the repos or the \fBAUR\fR\&.
.RE
.PP
yay -Si \fIfoo\fR
.RS 4
Gets information about package \fIfoo\fR from the repos or the \fBAUR\fR\&.
.RE
.PP
yay -S \fIfoo\fR --mflags "--skipchecksums --skippgpcheck"
.RS 4
Installs \fIfoo\fR while skipping checksums and pgp checks\&.
.RE
.PP
yay --devel --save
.RS 4
Sets devel to true in the config\&.
.RE
.PP
yay --stats
.RS 4
Shows statistics for installed packages and system health\&.
.RE
.SH "FILES"
.sp
\fBCONFIG DIRECTORY\fR
.RS 4
The config directory is \fI$XDG_CONFIG_HOME/yay/\fR\&. if
\fB$XDG_CONFIG_HOME\fR is unset, the config directory will fall back to
\fI$HOME/.config/yay\fR\%.
.PP
\fIconfig.json\fR\& Is used to store all of Yay's config options\&. Editing
this file should be done through Yay, using the options
mentioned in \fBPERMANENT CONFIGURATION SETTINGS\fR\&.
.RE
.PP
\fBCACHE DIRECTORY\fR
.RS 4
The cache directory is \fI$XDG_CACHE_HOME/yay/\fR\&. if
\fB$XDG_CACHE_HOME\fR is unset, the cache directory will fall back to
\fI$HOME/.cache/yay\fR\&.
.PP
\fIaur_<shellname>\fR holds a list of of all packages, including the AUR,
for shell completion\&. The completion files are refreshed every 48 hours\&.
.PP
\fIvcs.json\fR tracks VCS packages and the latest commit of each source. If
any of these commits change the package will be upgraded during a devel update.
.RE
.PP
\fBBUILD DIRECTORY\fR
.RS 4
Unless otherwise set this should be the same as \fBCACHE DIRECTORY\fR. This
directory is used to store downloaded AUR Packages as well as any source files
and built packages from  those packages\&.
.RE
.PP
\fBPACMAN.CONF\fR
.RS 4
Yay uses Pacman's config file to set certain pacman options either through
go-alpm or Yay itself. Options inherited include most libalpm options and
pacman options\&.
.PP
Notably \fBDatabases\fR, \fBColor\fR and \fB*Path/*Dir\fR options are used\&.
.RE
.PP
.SH "SEE ALSO"
.sp
\fBmakepkg\fR(8)
\fBPKGBUILD\fR(5)
\fBpacman\fR(8)
\fBpacman\&.conf\fR(5)
.PP
See the arch wiki at https://wiki\&.archlinux\&.org/index\&.php/Arch_User_Repository for more info on the \fBAUR\fR\&.
.SH "BUGS"
.PP
Please report bugs to our GitHub page https://github\&.com/Jguer/yay
.SH "AUTHORS"
.sp
Jguer <joaogg3@gmail\&.com>
.br
Morgana <morganamilo@gmail\&.com>


yay.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/jguer/yay/aur"
	vcs "github.com/jguer/yay/aur/vcs"
	"github.com/jguer/yay/config"
	pac "github.com/jguer/yay/pacman"
)

func usage() {
	fmt.Println(`usage:  yay <operation> [...]
    operations:
    yay {-h --help}
    yay {-V --version}
    yay {-D --database} <options> <package(s)>
    yay {-F --files}    [options] [package(s)]
    yay {-Q --query}    [options] [package(s)]
    yay {-R --remove}   [options] <package(s)>
    yay {-S --sync}     [options] [package(s)]
    yay {-T --deptest}  [options] [package(s)]
    yay {-U --upgrade}  [options] <file(s)>

    New operations:
    yay -Qstats          displays system information
    yay -Cd              remove unneeded dependencies
    yay -G [package(s)]  get pkgbuild from ABS or AUR

    New options:
    --topdown            shows repository's packages first and then aur's
    --bottomup           shows aur's packages first and then repository's
    --noconfirm          skip user input on package install
	--devel			     Check -git/-svn/-hg development version
	--nodevel			 Disable development version checking
`)
}

var version = "2.116"

func parser() (op string, options []string, packages []string, changedConfig bool, err error) {
	if len(os.Args) < 2 {
		err = fmt.Errorf("no operation specified")
		return
	}
	changedConfig = false
	op = "yogurt"

	for _, arg := range os.Args[1:] {
		if arg[0] == '-' && arg[1] != '-' {
			switch arg {
			default:
				op = arg
			}
			continue
		}

		if arg[0] == '-' && arg[1] == '-' {
			changedConfig = true
			switch arg {
			case "--printconfig":
				fmt.Printf("%+v", config.YayConf)
				os.Exit(0)
			case "--gendb":
				err = aur.CreateDevelDB()
				if err != nil {
					fmt.Println(err)
				}
				err = vcs.SaveBranchInfo()
				if err != nil {
					fmt.Println(err)
				}
				os.Exit(0)
			case "--devel":
				config.YayConf.Devel = true
			case "--nodevel":
				config.YayConf.Devel = false
			case "--timeupdate":
				config.YayConf.TimeUpdate = true
			case "--notimeupdate":
				config.YayConf.TimeUpdate = false
			case "--topdown":
				config.YayConf.SortMode = config.TopDown
			case "--complete":
				config.YayConf.Shell = "sh"
				_ = complete()
				os.Exit(0)
			case "--fcomplete":
				config.YayConf.Shell = "fish"
				_ = complete()
				os.Exit(0)
			case "--help":
				usage()
				os.Exit(0)
			case "--noconfirm":
				config.YayConf.NoConfirm = true
				fallthrough
			default:
				options = append(options, arg)
			}
			continue
		}
		packages = append(packages, arg)
	}
	return
}

func main() {
	op, options, pkgs, changedConfig, err := parser()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	switch op {
	case "-Cd":
		err = cleanDependencies(pkgs)
	case "-G":
		for _, pkg := range pkgs {
			err = getPkgbuild(pkg)
			if err != nil {
				fmt.Println(pkg+":", err)
			}
		}
	case "-Qstats":
		err = localStatistics(version)
	case "-Ss", "-Ssq", "-Sqs":
		if op == "-Ss" {
			config.YayConf.SearchMode = config.Detailed
		} else {
			config.YayConf.SearchMode = config.Minimal
		}

		if pkgs != nil {
			err = syncSearch(pkgs)
		}
	case "-S":
		err = install(pkgs, options)
	case "-Syu", "-Suy", "-Su":
		if strings.Contains(op, "y") {
			err = config.PassToPacman("-Sy", nil, nil)
			if err != nil {
				break
			}
		}
		err = upgradePkgs(options)
	case "-Si":
		err = syncInfo(pkgs, options)
	case "yogurt":
		config.YayConf.SearchMode = config.NumberMenu

		if pkgs != nil {
			err = numberMenu(pkgs, options)
		}
	default:
		if op[0] == 'R' {
			vcs.RemovePackage(pkgs)
		}
		err = config.PassToPacman(op, pkgs, options)
	}

	var erra error
	if vcs.Updated {
		erra = vcs.SaveBranchInfo()
		if erra != nil {
			fmt.Println(err)
		}

	}

	if changedConfig {
		erra = config.SaveConfig()
		if erra != nil {
			fmt.Println(err)
		}

	}

	erra = config.AlpmHandle.Release()
	if erra != nil {
		fmt.Println(err)
	}

	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

// NumberMenu presents a CLI for selecting packages to install.
func numberMenu(pkgS []string, flags []string) (err error) {
	var num int

	aq, err := aur.NarrowSearch(pkgS, true)
	if err != nil {
		fmt.Println("Error during AUR search:", err)
	}
	numaq := len(aq)
	pq, numpq, err := pac.Search(pkgS)
	if err != nil {
		return
	}

	if numpq == 0 && numaq == 0 {
		return fmt.Errorf("no packages match search")
	}

	if config.YayConf.SortMode == config.BottomUp {
		printAURSearch(aq, numpq)
		pq.PrintSearch()
	} else {
		pq.PrintSearch()
		printAURSearch(aq, numpq)
	}

	fmt.Printf("\x1b[32m%s\x1b[0m\nNumbers: ", "Type numbers to install. Separate each number with a space.")
	reader := bufio.NewReader(os.Stdin)
	numberBuf, overflow, err := reader.ReadLine()
	if err != nil || overflow {
		fmt.Println(err)
		return
	}

	numberString := string(numberBuf)
	var aurInstall []string
	var repoInstall []string
	result := strings.Fields(numberString)
	for _, numS := range result {
		num, err = strconv.Atoi(numS)
		if err != nil {
			continue
		}

		// Install package
		if num > numaq+numpq-1 || num < 0 {
			continue
		} else if num > numpq-1 {
			if config.YayConf.SortMode == config.BottomUp {
				aurInstall = append(aurInstall, aq[numaq+numpq-num-1].Name)
			} else {
				aurInstall = append(aurInstall, aq[num-numpq].Name)
			}
		} else {
			if config.YayConf.SortMode == config.BottomUp {
				repoInstall = append(repoInstall, pq[numpq-num-1].Name())
			} else {
				repoInstall = append(repoInstall, pq[num].Name())
			}
		}
	}

	if len(repoInstall) != 0 {
		err = config.PassToPacman("-S", repoInstall, flags)
	}

	if len(aurInstall) != 0 {
		err = aur.Install(aurInstall, flags)
	}

	return err
}

// Complete provides completion info for shells
func complete() (err error) {
	path := os.Getenv("HOME") + "/.cache/yay/aur_" + config.YayConf.Shell + ".cache"

	if info, err := os.Stat(path); os.IsNotExist(err) || time.Since(info.ModTime()).Hours() > 48 {
		os.MkdirAll(os.Getenv("HOME")+"/.cache/yay/", 0755)

		out, err := os.Create(path)
		if err != nil {
			return err
		}

		if aur.CreateAURList(out) != nil {
			defer os.Remove(path)
		}
		err = pac.CreatePackageList(out)

		out.Close()
		return err
	}

	in, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0755)
	if err != nil {
		return err
	}
	defer in.Close()

	_, err = io.Copy(os.Stdout, in)
	return err
}
